Unitary system now reports warning when cycling supply fan schedule is used with setpoint control type

matthew-larson opened the pull request:
Pull request overview
---------------------
 - Fixes #8285 
 - This pull request corrects a check for the unitary system based on the Supply Air Fan Operating Mode Schedule when the Control Type is set to Setpoint.  Previously, the fan operation mode schedule was not being checked, which was keeping the cooling coil from being enabled.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

matthew-larson committed:
Add if statement for SolveRoot method and PartLoadRatio calculation.

matthew-larson committed:
Revert code change and add unit test for issue.

matthew-larson committed:
Merge branch 'develop' of https://github.com/nrel/energyplus into unitary-clgcoil-cycling

matthew-larson committed:
Modify fix location and unit test.

matthew-larson committed:
Merge branch 'develop' of https://github.com/nrel/energyplus into unitary-clgcoil-cycling

matthew-larson committed:
Merge branch 'develop' of https://github.com/nrel/energyplus into unitary-clgcoil-cycling

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -7996,6 +7996,15 @@ namespace UnitarySystems {
+                        DataHVACGlobals::OnOffFanPartLoadFraction = 1.0;
+                    }
+                }
+
```
Well, this works but now this same code is in 2 places, here and in initLoadBasedControl. Shouldn't these get moved to initUnitarySystems inside the FirstHVACIteration test (line 1084) so that both control types execute this, it's only done once per time step and there is only 1 copy of these lines of code.

matthew-larson replied:
That makes sense, I'll make this adjustment once the CI finishes.  Thanks

matthew-larson committed:
Merge branch 'develop' of https://github.com/nrel/energyplus into unitary-clgcoil-cycling

matthew-larson committed:
Merge fan operating mode if statement into InitUnitarySystems.

matthew-larson committed:
Merge branch 'develop' of https://github.com/nrel/energyplus into unitary-clgcoil-cycling

matthew-larson commented:
@rraustad I moved the fix to initUnitarySystems as recommended to eliminate the duplicated check.  If you could make sure I got it in the right spot, it's ready for another review.

rraustad commented on tst/EnergyPlus/unit/UnitarySystem.unit.cc:
```diff
@@ -15406,3 +15406,234 @@ TEST_F(ZoneUnitarySysTest, UnitarySystemModel_MultiSpeedDXCoilsDirectSolutionTes
+    EXPECT_NEAR(DataLoopNode::Node(2).Temp, DataLoopNode::Node(2).TempSetPoint, 0.001);
+    // cooling coil air inlet node temp is greater than cooling coil air outlet node temp
+    EXPECT_GT(DataLoopNode::Node(3).Temp, DataLoopNode::Node(2).Temp);
+}
```
Shoot, this was ready to merge. Line feed needed at end of unit test.

rraustad reviewed (changes requested):
Revision to code location so load based and set point based control use same logic. This looks good. Unit test still needs mod to last line.

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -991,6 +991,15 @@ namespace UnitarySystems {
             DataLoopNode::Node(this->m_HeatRecoveryInletNodeNum).MassFlowRate = mdotHR;
         }
 
+        if (this->m_FanOpModeSchedPtr > 0) {
```
Well, I actually suggested this be moved to line 1006, or better yet 1010, below since it needs only be done once per time step. Might as well make that change while you correct the unit test.

matthew-larson committed:
Move code fix to first iteration only and add line feed and end of unit test.

rraustad commented:
Perfect, I think I'll run the defect file before merging this. Also, check off the pull request author items in the pull request description. Think about the PR title from the users perspective.

matthew-larson committed:
Fix unit test from develop changes.

rraustad commented:
Thanks @matthew-larson. You are done with this one unless something else unexpectedly pops up. I'll get to testing the defect file when I can.

matthew-larson commented:
Sounds good, thanks for your help and walking me through the code, it was very valuable.  I'll keep tabs on this to make sure everything comes back clean.

matthew-larson committed:
Merge branch 'develop' of https://github.com/nrel/energyplus into unitary-clgcoil-cycling

matthew-larson committed:
Change fix to provide correct error message for cycling fan with setpoint control.

matthew-larson committed:
Update unit test for error checking of cycling fan with setpoint control.

matthew-larson commented:
@rraustad I updated the pull request for the new fix of making sure the correct error is caught when a cycling fan is used with Setpoint control.  I also updated the unit test to check for the error.  Let me know if you have any additional comments on this.

rraustad commented on tst/EnergyPlus/unit/UnitarySystem.unit.cc:
```diff
@@ -15406,3 +15406,159 @@ TEST_F(ZoneUnitarySysTest, UnitarySystemModel_MultiSpeedDXCoilsDirectSolutionTes
+    DataZoneEquipment::ZoneEquipInputsFilled = true;                                    // indicate zone data is available
+    thisSys->getUnitarySystemInputData(state, compName, zoneEquipment, 0, ErrorsFound); // get UnitarySystem input from object above
+    std::string const error_string = delimited_string({
+        "   ** Severe  ** AirLoopHVAC:UnitarySystem = UNITARY SYSTEM MODEL\n   **   ~~~   ** For FAN:ONOFF = SUPPLY FAN 1\n   **   ~~~   ** Fan operating mode must be continuous (fan operating mode schedule values > 0).\n   **   ~~~   ** Error found in Supply Air Fan Operating Mode Schedule Name ALWAYSZERO\n   **   ~~~   ** ...schedule values must be (>0., <=1.)"
```
@matthew-larson are you clang formatting as you go? It seems this line would have wrapped. Not sure if it's worth another push if everything else is OK.

matthew-larson replied:
I have clang formatting ON in CLion.  I can double check that it's working right.

rraustad commented:
This seems correct now.

The DOAS unit uses setpoint control and FanOpMode schedule value = 0

    ** Severe  ** AirLoopHVAC:UnitarySystem = DOAS UNITARY
    **   ~~~   ** For FAN:ONOFF = DOAS UNITARY SUPPLY FAN
    **   ~~~   ** Fan operating mode must be continuous (fan operating mode schedule values > 0).
    **   ~~~   ** Error found in Supply Air Fan Operating Mode Schedule Name DOAS UNITARY FAN MODE SCHEDULE
    **   ~~~   ** ...schedule values must be (>0., <=1.)
    **  Fatal  ** getUnitarySystemInputData: previous errors cause termination. Check inputs

If value for FanOpMode schedule is changed to 1 or that input is left blank then there are no fan op mode schedule warnings and the coil turns on:

![image](https://user-images.githubusercontent.com/5778156/97894948-24880980-1d01-11eb-908a-bbcba2e756fd.png)

The check of the FanOpMode schedule for both set point control and load or SZVAV control is not necessary so I'm glad that code was not ultimately moved as discussed. For set point control, we know the fan op mode must be constant so only checking once that the schedule value is > 0 is the only thing needed.

rraustad commented:
Running unit tests locally multiple times to get a clean pass. Keeps stalling at LoggingSubStep4stepsPerHour. That unit test should not be related to this change.

mitchute commented:
> Keeps stalling at LoggingSubStep4stepsPerHour

@rraustad, FWIW, that test usually takes a few minutes to complete on my machine.

rraustad commented:
Yeah, that unit test finally finished after 15 minutes, using a release build.

![image](https://user-images.githubusercontent.com/5778156/97904904-2bb61400-1d0f-11eb-87ff-61c936bc5dd3.png)

rraustad reviewed (approved):
Simple changes made to catch issues with fan operating mode schedule when set point control is used.

mitchute commented:
15 minutes... 😮
